File: central/laboratory/avalon-net/Composition/ModelDefault/SimpleFIFO.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/

namespace Apache.Avalon.Composition.Model.Default
{
	using System;
	
	/// <summary>This class implements a simple thread-safe FirstIn-FirstOut queue.
	/// 
	/// It is not intended to be used outside this package,
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.2 $ $Date: 2004/03/07 22:06:40 $
	/// </version>
	class SimpleFIFO
	{
		private System.Collections.ArrayList m_Queue;
		
		internal SimpleFIFO()
		{
			m_Queue = new System.Collections.ArrayList();
		}
		
		public void Clear()
		{
			lock (this)
			{
				m_Queue.Clear();
			}
		}
		
		public void Put(System.Object obj)
		{
			lock (this)
			{
				m_Queue.Add(obj);
				System.Threading.Monitor.PulseAll(this);
			}
		}
		
		public System.Object Get()
		{
			lock (this)
			{
				// while (m_Queue.Count == 0)
				//	System.Threading.Monitor.Wait(this, TimeSpan.FromMilliseconds(100));
				if (m_Queue.Count == 0)
				{
					return null;
				}
				object ret = m_Queue[0];
				m_Queue.RemoveAt(0);
				return ret;
			}
		}
		
		public int Size
		{
			get
			{
				lock (this)
				{
					return m_Queue.Count;
				}
			}
		}
	}
}

File: avalon/activation/impl/src/java/org/apache/avalon/activation/af4/SimpleFIFO.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.activation.af4;

import java.util.ArrayList;

/** This class implements a simple thread-safe FirstIn-FirstOut queue.
 * 
 * It is not intended to be used outside this package,
 * 
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.1.2.1 $ $Date: 2004/02/22 15:50:06 $
*/
class SimpleFIFO
{
    private ArrayList m_Queue;
    
    SimpleFIFO()
    {
        m_Queue = new ArrayList();
    }
    
    void clear()
    {
        synchronized( this )
        {
            m_Queue.clear();
        }
    }
    
    void put( Object obj )
    {
        synchronized( this )
        {
            m_Queue.add( obj );
            notifyAll();
        }
    }
    
    Object get()
        throws InterruptedException
    {
        synchronized( this )
        {
            while( m_Queue.size() == 0 )
                wait(100);
            return m_Queue.remove(0);
        }
    }
    
    int size()
    {
        synchronized( this )
        {
            return m_Queue.size();
        }
    }
} 
